Connections ui test shot.

leamas opened the pull request:
This is not really a PR in the sense it should be merged. It's more about a demo.

It implements my initial ideas in #4078. The important thing is a new infrastructure to provide driver feedback. This makes it possible to create a more responsive UI where the state of the connection is directly visible.

The thing is available using the menubar, Tools | Connections. Sample screenshot:

![image](https://github.com/user-attachments/assets/709735fd-692c-4427-84dd-088a494659ff)

More thinking is needed when comparing to the current GUI below. Not least, the current view presents some rather obscure options like  *Use magnetic bearings in output sentence APB* in first panel . Overall using a cleaner interface like this PR for the connections should probably come with a better way to handle these overall communication parameters:

![image](https://github.com/user-attachments/assets/5527d83c-82dd-4f0d-9177-52dc3d89b16a)

leamas commented:
Notes:
  - I have been sloppy with the utf-8 characters in the UI - these probably does not work on Windows as of now.
  -  The driver statistics is currently only implemented for serial N0183 and network N0183.

rgleason commented:
Are menus changing?  What menu does the nmea debug window land in?
I guess we will need changes to the manual too.

leamas commented:
See #4078

That said, it's a long way before we can see possible manual changes.

bc38 commented:
Looks great!
Why did you abandon the traffic light status?
Maybe we should add icons from OpenBridge?
I suggest adding icons to the table for setting priorities

leamas commented:
> Looks great! Why did you abandon the traffic light status? Maybe we should add icons from OpenBridge? I suggest adding icons to the table for setting priorities


I abandoned the traffic lights just because the looked bad. Modern guls don't use colours this way, it does not blend well w.ith the rest IMHO 

Here are no icons, just utf codepoints. This has many advantages such as being well known from other contexts and also hopefully working out of the box in different colour schemes

As for the priorities I need to limit this to the overview. For now I don't intend to make any changes at all to the rest. If I land this there will be opportunities for new such PR

Thanks for feedback!

leamas commented:
Pushed a new version. screenshot:

![image](https://github.com/user-attachments/assets/87c202e4-1443-4c4f-b5cd-f1567a57a306)

This is aimed to be an almost complete UI. While the table overview is reasonably  functional all the hooks for the rest (checkboxes etc.) are empty. The goal is to finalize the GUI first.

Need to think more about Håkan's filter column. I am quite convinced that we cannot mention the word "filter" in the overview by default since these for good reasons are hidden under Advanced.  But, we could perhaps have a column labeled "Notes" or so. This could get filter info (icons) if/when they are defined, and would also be open for possible other "notes".

Input welcome, especially before the GUI gets finalized...

Hakansv commented:
On W10
Generally very good but some quick comments. More may come but probably not today. :)
See my shot for reference:
1. Port: User comments are now  there but instead of the address and port.  The address:port is for me first hand info and the comment nice to have. Could the comment come to a second row? May need a vertical ruler?
2. Status: The "Not used filled circle" is rather dominant. The previous unfilled circle may be an alternative ?
3. Status: The "!" sign means No problem - no traffic. Correct?
4. No Column header neither for "Edit" nor "Delete". Thoughts?
5. Delete mark: You see the difference from your shot and mine? I'd like the same as your shot.
6. General sizing. My shot is the whole screen height. From your shot there should be the prioritization button. Does not fit here.
7. General dialog window: Now it's hidden behind OCPN as soon I click the canvas. Compare to the windows function for e.g. "Dashboard" or "NMEA Debug window" that's always on top of OCPN and visible. Could that function be copied?

![bild](https://github.com/user-attachments/assets/dbd7266b-007f-41b3-afd6-087db4586988)

leamas commented:
> Port: User comments are now there but instead of the address and port. The address:port is for me first hand info and the comment nice to have. Could the comment come to a second row? May need a vertical ruler?

No. The added complexity is not worth it. Less is more.

 Another solution might to have the Description/User commetn available as a tooltip. Thoughts on that?

> Status: The "Not used filled circle" is rather dominant. The previous unfilled circle may be an alternative 

They have different meaning. The unfilled means "No stats available" whereas the filled is "Disabled".  The tooltips informs about it.  Have you ideas for another utf codepoint I'm open for that.

> Status: The "!" sign means No problem - no traffic. Correct?
No Column header neither for "Edit" nor "Delete". Thoughts?

Since there are correct tooltips in all cases  and the edit/dekete icons are well known: I sthi actually a problem

> General sizing. My shot is the whole screen height. From your shot there should be the prioritization button. Does not fit here.

Yes, you have an interesting use case with good know how many connections ;) The solution is of course to use a scrolled window so the connections view is limited, leaving place for the other stuff.

> General dialog window: Now it's hidden behind OCPN as soon I click the canvas. Compare to the windows function for e.g. "Dashboard" or "NMEA Debug window" that's always on top of OCPN and visible. Could that function be copied?

Eventually, this will replace the current view under Options;  no point trying to sort this out until then

leamas commented:
> Delete mark: You see the difference from your shot and mine? I'd like the same as your shot.

This is about windows utf support, in particular utf-8. Here, win32 is actually a pain. That said, there must be ways to render utf codepoints with more than 16 bits also on Windows. This is the difference, all codepoints besides delete   are 16 bits.

Might replace with  crossmark: ❌ (#274c) on windows to get around the OS-related mess for now.

Hakansv commented:
> > Delete mark: You see the difference from your shot and mine? I'd like the same as your shot.
> 
> This is about windows utf support, in particular utf-8. Here, win32 is actually a pain. That said, there must be ways to render utf codepoints with more than 16 bits also on Windows. This is the difference, all codepoints besides delete are 16 bits.
> 
> Might replace with crossmark: ❌ (#274c) on windows to get around the OS-related mess for now.

Yes, good for me.

Hakansv commented:
> > Port: User comments are now there but instead of the address and port. The address:port is for me first hand info and the comment nice to have. Could the comment come to a second row? May need a vertical ruler?
> 
> No. The added complexity is not worth it. Less is more.

I don't understand. I think we should show address and port in first hand.

> 
> Another solution might to have the Description/User comment available as a tooltip. Thoughts on that?
> 
Yes OK but I've no tooltips whatsoever?

> > Status: The "Not used filled circle" is rather dominant. The previous unfilled circle may be an alternative
> 
> They have different meaning. The unfilled means "No stats available" whereas the filled is "Disabled". The tooltips informs about it. Have you ideas for another utf codepoint I'm open for that.
> 

As said, no tooltips seen.

> > Status: The "!" sign means No problem - no traffic. Correct?
> > No Column header neither for "Edit" nor "Delete". Thoughts?
> 
> Since there are correct tooltips in all cases and the edit/delete icons are well known: Is this actually a problem?
> 

No with working tooltips it's OK.

> > General sizing. My shot is the whole screen height. From your shot there should be the prioritization button. Does not fit here.
> 
> Yes, you have an interesting use case with God knows how many connections ;) The solution is of course to use a scrolled window so the connections view is limited, leaving place for the other stuff.
> 

Yes, I'm not the last idiot. And we have no restrictions in connections so this can happen also "out there".

> > General dialog window: Now it's hidden behind OCPN as soon I click the canvas. Compare to the windows function for e.g. "Dashboard" or "NMEA Debug window" that's always on top of OCPN and visible. Could that function be copied?
> 
> Eventually, this will replace the current view under Options; no point trying to sort this out until then

OK.

Hakansv commented:
re: "Might replace with crossmark: ❌ (#274c) on windows "
As said OK but why is the gear icon for "Edit" shown but not the litter box from your shot? Any difference in coding?

leamas commented:
> Yes OK but I've no tooltips whatsoever?

This needs some investigation. It works for me, and I don't really understand why it should be platform dependent. But you could never know, can you?

leamas commented:
> Yes, I'm not the last idiot. And we have no restrictions in connections so this can happen also "out there".

Don't get me wrong, your use  case is valid  and we definitely need to handle it. Furthermore, it's a good testcase pushing some things to their limits.

Hakansv commented:
> > Yes OK but I've no tooltips whatsoever?
> 
> This needs some investigation. It works for me, and I don't really understand why it should be platform dependent. But you could never know, can you?

No, but I know the tool bar's tooltips are working. You may use another method? 
btw: it's a similar experience for PR #4051

Hakansv commented:
Other observations while "on it"

"Filter NMEA Course and speed..."  settings are not shown, see shot.

Some strings in connections_dlg.cpp are not marked for translation although I suppose it should?
I've found it on rows: 106, 143, 145, 312(?)

I tried the "Edit.." and "Delete.." functions but a click on either of it don't work.  (My thought was to test the "Ok to delete connection on " but nothing happens. On last commits a test of delete worked but not now.) 


Settings are not shown:
![bild](https://github.com/user-attachments/assets/08b96a80-bc06-40d5-b55c-cd8ae0b643a4)

leamas commented:
Pushed a new version. Hopefully fixed issues:

  - The Trashcan icon i. e. using 32-bits utf codepoints finally sorted out by Dan (thanks!). The correct icon should now be visible also on Windows.  This also means  that we can get rid of the awful wxFormat hack used in multiple places to handle the degrees symbol.
  - The connection grid now lives in a scrollable area if/when there is not space for all lines.
  - The delete function is fixed.
  - Various polish

Overall state:
  - The grid with connections should be feature complete, including the + sign.
  - Remaining panels (General, Advanced) are basically just mock ups. There is no use testing the functionality, the question is how it looks.
  - I see multiple needs for UI enhancements, in particular sizing issues caused by the scrolled window.
  - One major bug: missing tool tips on Windows.

Windows missing tooltips: I need some help debugging this.  There is really two questions, both related to the following snippet from connections_dlg.cpp around line 116:
```
 GetGridWindow()->Bind(wxEVT_MOTION, [&](wxMouseEvent ev) {
      wxPoint pt = ev.GetPosition();
      int row = YToRow(pt.y);
      int col = XToCol(pt.x);
      if (col < 7 && row < GetNumberRows()) 
         SetToolTip(m_tooltips[row][col]);
      ev.Skip();
    });
```
First question:  is this lambda is triggered at all? That is, if there is a breakpoint at `wxPoint pt = ev.GetPosition()`, is this hit as soon as the mouse is moved over the grid?

Second question, if answer to above is "yes":  Is the line `SetToolTip(m_tooltips[row][col]);` hit if moving the mouse over the grid? Please note I have broken the original line into two to make it easier to set a breakpoint.

@Hakansv Unfortunately, the links are not that useful, they are both about corner cases. However, this is not a corner case in any sense, it's just a plain wxWidgets app.

rgleason commented:
Does 0000274C work to represent the Cross Mark emoji in windows 32 bits?
Yes, the UTF-32 representation 0000274C works to represent the Cross Mark emoji (❌) in a 32-bit system. Windows systems can interpret this 32-bit Unicode code point correctly, as long as the software or application you're using supports UTF-32 encoding. The key difference is that UTF-32 uses a fixed length of 4 bytes (32 bits) to represent each character, including emojis like the Cross Mark.

leamas commented:
@rgleason  Do you paste output from some AI here? If so, this is a quite nice example on the need to raise the correct question and to understand the answer. I this context, I fail to understand any of them.

tl;dr; The Unicode/utf-8/wxString mess is resolved.  Crossmark has never been an issue, trashcan has.

transmitterdan commented:
> Pushed a new version. Hopefully fixed issues:
> 
> First question: is this lambda triggered at all? That is, if there is a breakpoint at `wxPoint pt = ev.GetPosition()`, is this hit as soon as the mouse is moved over the grid?

Yes.

> 
> Second question, if answer to above is "yes": Is the line `SetToolTip(m_tooltips[row][col]);` hit if moving the mouse over the grid?
Yes, but the tooltip string it tries to set is "" (empty string).  I believe that rows and cols are zero based, not 1 based.  So, first row should be row 0.

Hakansv commented:
- Yes, trash bin is nice, very good.
- And "Connections" with the ruler is good. Now it can be sized to fit.
- Type:Address:port is back, good. But what about user comments?
- 
Issues:
- The Connections dialog can be resized but the new size and it's location is not saved so on next open I've to adjust size and place again.
- The New/Edit dialog is to small and can't be resized. I can't fold down to select data port. See shot. 
- Why a plus-icon instead of a push button for add connection? Personally, I don't think it fits in with the rest, but that's me.
- Unlike for example the old Option->Connections or Dashboard or .., the new Connection is hidden behind OCPN in case I switch windows or edit a connection. I think it's important that the placement is always above OCPN and always below other apps.

That's for now. I'll try to debug the tooltips as of you advice above.

![bild](https://github.com/user-attachments/assets/ce9b4dce-7a02-450e-9ec4-35a58f98c08b)

transmitterdan commented:
> > Pushed a new version. Hopefully fixed issues:
> Yes, but the tooltip string it tries to set is "" (empty string). I believe that rows and cols are zero based, not 1 based. So, first row should be row 0.

My apologies, I spoke too soon.  The string it tries to set is right.  Maybe the tooltip is not enabled for some reason?  The code here seems to do what you expected it to do.  I sets the tooltip string to the right string.

transmitterdan commented:
> > > Pushed a new version. Hopefully fixed issues:
> > > Yes, but the tooltip string it tries to set is "" (empty string). I believe that rows and cols are zero based, not 1 based. So, first row should be row 0.
> 
> My apologies, I spoke too soon. The string it tries to set is right. Maybe the tooltip is not enabled for some reason? The code here seems to do what you expected it to do. I sets the tooltip string to the right string.

The mistake is to set the text on the grid element.  You want to set the tooltip on the Grid window itself. Try changing this:
```
  SetToolTip(m_tooltips[row][col]);
```
to this:
```
  GetGridWindow->SetToolTip(m_tooltips[row][col]);
```

transmitterdan commented:
Here is an "improved" lambda:

```
    GetGridWindow()->Bind(wxEVT_MOTION, [&](wxMouseEvent ev) {
      static int oldRow = -1;
      static int oldCol = -1;  
      wxPoint pt = ev.GetPosition();
      int row = YToRow(pt.y);
      int col = XToCol(pt.x);
      if (row == wxNOT_FOUND || col == wxNOT_FOUND) return;
      if (col != oldCol || row != oldRow)
        GetGridWindow()->SetToolTip(m_tooltips[row][col]);
      ev.Skip();
      oldRow = row;
      oldCol = col;
    });
```

This avoids the constant repainting of the tooltip every time the mouse moves unless it actually changes grid position.

Hakansv commented:
@transmitterdan Lovely! Your patch is working.
And the question above about User comment is solved!

![bild](https://github.com/user-attachments/assets/c087fbe5-a583-4489-8c5f-b23a422c889f)

leamas commented:
The tooltip was the major thing to fix. As usual, in hindsight it was just a silly bug which for some reason was hidden on Linux. Thanks, Dan!

>  The Connections dialog can be resized but the new size and it's location is not saved so on next open I've to adjust size and place again.

Let's postpone this until the new dialog is merged into Options.  

> The New/Edit dialog is to small and can't be resized. I can't fold down to select data port. See shot.

This is not a new dialog, it's more or less verbatim the old one. And this is seriously broken, the layout mechanism does not work due to some design problems.. Currently, the  dialog has a fixed size, but having a fixed size is basically a bug. I could of course just use  the current size... (sigh)

 >  Why a plus-icon instead of a push button for add connection? Personally, I don't think it fits in with the rest, but that's me.

I actually agree.

 > Unlike for example the old Option->Connections or Dashboard or .., the new Connection is hidden behind OCPN in case I switch windows or edit a connection. I think it's important that the placement is always above OCPN and always below other apps.

Again, let's wait until  this is merged into Options. 

Another use case to consider is a new user where no connections are defined...

transmitterdan commented:
> @transmitterdan Lovely! Your patch is working. And the question above about User comment is solved!
> 
This needs testing a lot.  I think there are still corner cases where it might leave a dangling tool-tip.  Also, on Windows this dialog gets shoved behind OpenCPN canvas window pretty often.  Very annoying.

Hakansv commented:
> Again, let's wait until this is merged into Options.

Sorry, my memory is not bad just a sometimes short ...  :)

> Another use case to consider is a new user where no connections are defined...

Wouldn't it be just a empty list and the "Make New" button. If you do not show the "Connections dialog" if zero connections a "dummy" may be needed?

leamas commented:
> Here is an "improved" lambda:

hm... I'm basically skeptical about statics except in very rare cases like singleton constructors. This is not...

I have also encountered crashes when YToRow or XToCol returns values out of range when resizing the window fast. Hence, we need the current checks for valid row/col as they are.

Yes, currently the tooltip is redrawn  for each move event. But: is this actually a problem?  I suggest that we leave this without optimizations for now and makes test on Rpi before final conclusions.ö

transmitterdan commented:
On Windows it's really annoying.  Constantly flickering every mouse event which number in the hundreds per second.

I agree, this needs lots testing as I am sure there are more than a few corner cases.

leamas commented:
@transmitterdan : the plan is that teh new dialog should replace the current dialog in Optiosn | Connections. Once in place, it will behave like today when it comes to (not) staying other windows etc.

leamas commented:
> On Windows it's really annoying. Constantly flickering every mouse event which number in the hundreds per second.

OK, let's fix it then. But no statics.

transmitterdan commented:
Why use a lambda when it isn't really needed?

leamas commented:
Well. it's partly a question of style. When binding to an event like here, everything is in one place and the namespace isn't polluted. In my eyes, it's just better.

That said, lambdas should be short and this is stretching it. One or two more lines would make it go into a function. Me being me, I would still call it in a lambda, though.

Current status:
```
 GetGridWindow()->Bind(wxEVT_MOTION, [&](wxMouseEvent ev) {
      wxPoint pt = ev.GetPosition();
      int row = YToRow(pt.y);
      int col = XToCol(pt.x);
      if (col >= 0 && col < 7 && row >= 0 && row < GetNumberRows()) {
        auto new_tooltip = m_tooltips[row][col];
        if (new_tooltip != GetGridWindow()->GetToolTipText())
          GetGridWindow()->SetToolTip(new_tooltip);
      }
      ev.Skip();
```

leamas commented:
> But what about user comments?

available as tooltip for Port

transmitterdan commented:
> EDIT: you are in a way right, this deserves a separate function. Done by me.

I agree.  Ok, I stop here but this is the event handler function I created in case you are interested.  It handles all the corner cases I could think of and doesn't flash the tooltip. But is not "optimized".

```
private:
  int oldRow, oldCol;
  void onMouseMove(wxMouseEvent& ev) {
    wxPoint pt = ev.GetPosition();
    int row = YToRow(pt.y);
    int col = XToCol(pt.x);
    if (col >= 0 && row >= 0) {
      wxString toolText = m_tooltips[row][col];
      wxToolTip* toolTip = GetGridWindow()->GetToolTip();
      if (toolTip) {
        if (toolTip->GetTip() != toolText || oldRow != row || oldCol != col) {
          GetGridWindow()->SetToolTip(toolText);
          oldRow = row;
          oldCol = col;
        }
      } else {
        GetGridWindow()->SetToolTip(toolText);
        oldRow = row;
        oldCol = col;
      }
    } else
      GetGridWindow()->UnsetToolTip();
    ev.Skip();
  }
```

leamas commented:
what if you test my one above which is shorter and also checks for too big col/row?

leamas commented:
... otherwise, there will be a new upload later...

Hakansv commented:
> what if you test my one above which is shorter and also checks for too big col/row?

I made a check on that as well..
A dated, slow, Win10 and screen 14" 2560:1440 scaled 175%
The tooltips are working. Though a strange effect on col 6(edit) and 7(Delete):
If hoover e.g. row 1, col 7 and move the mouse straight down to row 2, 3, 4...  the tooltip is not replaced and moved but  stays at row 1. Not happening on col 4 (port). but that could be since the text changes?

leamas commented:
New upload. Needs fixes, nothing to test

leamas commented:
New upload.

Major change: The ui has moved to Options | Connections, the old menu entry Tools | Connections is gone.

Fixes:
  - The tooltips should be in place and working.
  - The edit connections dialog now has the original, fixed size. The fixed size is a bug, but this requires separate work.
  - The + sign is replaced by a button.
  - Enhanced handling of no connections at all, avoiding displaying an empty table.
  - Lots of polish.

Issues:

  - On windows, the grid is for some reason equipped with a black background making the text unreadable for me. Can be worked around using another colour scheme for now.
  - I have not implemented the sort order stuff. Tend to think that the "Sort order" configuration should be removed. If sorting is needed it's better to implement the ubiquitous "click on a column header" sorting.
  - It should be possible to let the "Add connection" button to stay attached to the grid when dialog is expanded. I have so far failed. though.

leamas commented:
New screenshot: 
![image](https://github.com/user-attachments/assets/e24364fb-99c8-4d48-bc93-cb3f4cf84662)

Hakansv commented:
Nice! I'll test soon. Other stuff in between...
On some "issues":

> I have not implemented the sort order stuff. Tend to think that the "Sort order" configuration should be removed. If sorting is needed it's better to implement the ubiquitous "click on a column header" sorting.

I agree. The sort order is a leftover from when the present source prioritization was implemented. It was simply changed to list order instead of the former simple N0183 sentence prioritization. I think no real need anymore. Does actually confuse some users.

>  Is there a valid use case for two connections with the same (bus, port) tuple?

Yes could be. My example is:
1. TCP 192.168.1.78:10110  - No filter
2. TCP 192.168.1.78:10110  - Filter to only receive VDM

The important thing is to keep the watch-out for input and output of the same data without filter. To avoid a feedback loop.

Hakansv commented:
Thoughts on sort order:
I think you can find a recent way.
One can be to sort by "type" OR "enabled" OR input - in/out - output ....

leamas commented:
No, I discovered that. But latest upload does.

transmitterdan commented:
Also, if you manually create the grid cells you need to choose a background color attribute (default is NULL or black) for each cell. Letting if fall to the default raises an exception in Debug builds. And the background and text colors need to be coordinated with the day/dusk/night global color scheme so letting it default isn't probably what we want anyway. See how StatusBar or other parts of the Options Dialog does it for example.

Hakansv commented:
I think you've to rebase to current master? 
Conflicts in model/src/comm_util.cpp

leamas commented:
>  think you've to rebase to current master?
Conflicts in model/src/comm_util.cpp

I have already done this here. However, I don't want to push too often, might wear down others ;)

If you want to test (appreciated!) please use this branch as it is right now. Otherwise, just wait for next upload. I'm working on the open issues [above](https://github.com/OpenCPN/OpenCPN/pull/4283#issuecomment-2573778324), it might take some time.

transmitterdan commented:
Just trying to help: This mostly makes the wxGrid in the new dialog follow the day/dusk/night color scheme about as well as other dialogs.  Most dialogs don't get everything right.  Would be nice to work on the general case as many users complain about it.
[connections_dlg.txt](https://github.com/user-attachments/files/18331627/connections_dlg.txt)

Hakansv commented:
> If you want to test (appreciated!) please use this branch as it is right now.

I think I fixed the conflict but got other build errors. I'll await your update.
Thanks

leamas commented:
@transmitterdan :+1: 

Applied your colour scheme patch. Works like a much needed crash course for me. Thanks!

On Ĺinux, most things looks good.  However, using the default scheme it looks like below. Is this as it should be? From a pure aesthetic  point of view I think it was better before the patch where all widgets had the same light background.

Also, when looking into `void DimeControl()` in  *navutil.cpp* I see a lot of exceptions for Macos (and for _Win32). 

Thoughts?

![image](https://github.com/user-attachments/assets/fa566a70-08fd-49d6-98ee-423cfe9f2fcf)

leamas commented:
New upload, feature complete.  

Fixed issues:
  -  The black background in Windows  should be fixed (thanks, Dan)
  -  Sorting is implemented. Click on column header to sort w r t that  column.
  -  The OK/Cancel/Apply is fully implemented (might very well be the only one doing this...)
  -  Connections and drivers are connected using an unique key, similar but not the same as the serialized format. Should handle multiple connections with same (bus, iface) tuple. Here might very well be dragons, for sure.

Known issues:
  - The colour scheme handling needs to be stabilized, see [above]()
  - The placement of the "Add new connection" button still haunts me.

Hakansv commented:
You are working hard!
Sorry not so good on Win. Too many ambiguities to list tonight.

The shot shows what's there. Nothing else.
The check boxes to enable/disable are though working. But need two clicks to react.
The "Edit" buttons are also working but no good dialog, shot 2. Not possible to resize.
The sort function is working and a good philosophy.

![bild](https://github.com/user-attachments/assets/649dc8d4-1142-41ba-8ae7-d471619da6e7)

![bild](https://github.com/user-attachments/assets/5da34c4a-6011-432b-be8a-c94f59489535)

leamas commented:
The edit dialog is the old one,  more or less verbatim (besides that I missed to remove the list position combox) It has never been possible to resize, the question is just how the fixed size should be.

It's strange that you don't get a scrollable area instead of all connections at once. When I tested on windows I got a really small scrollable window, way too small. This is not only about windows, it's about your test case.

transmitterdan commented:
I think would be worth it to fix dime control to apply all needed coloring of various controls. It's either that or every control needs to self manage colors of all elements. For example, the grid dividing lines need to be color managed as well as the automatic scroll bars. It's a daunting task. So much so that I wonder whether wxWidgets has something. But as I understand it this was originally the purpose of dime control. It just hasn't been maintained so it handles all the derived wxWidgets controls like wxGrid.

I sort of hate all the OS specific coloring. Makes maintenance and testing harder. If we are going to do that I would rather do it in the color lookup function than in the dialog code. If a color is suboptimal for one dialog it's also same in any dialog that uses that wx control. So we could add lookups for wxGrid background and text colors same as regular text box background and foreground colors. Basically, expand the purview of the color looker-upper.

transmitterdan commented:
I think using default BG color can never work because it is then not able to follow the day/dusk/night/dawn color schemes. So. using default is a non-starter I think. And I think it is actually an error in wxWidgets mind.  At least on Windows it generates an error trap in DEBUG build to alert the tester something is wrong.

leamas commented:
> I think using default BG color can never work because it is then not able to follow the day/dusk/night/dawn color schemes. 

However, it does on Linux: 
![image](https://github.com/user-attachments/assets/06ba042e-5908-4b17-b70b-da5817b1308a)

I have not looked into the details, but my overall understanding is that on Linux and MacOS the colour schemes  control is used to  configure the application wide theming.  

> I think would be worth it to fix dime control to apply all needed coloring of various controls.

+1

> I sort of hate all the OS specific coloring. Makes maintenance and testing harder. If we are going to do that I would rather do it in the color lookup function than in the dialog code.

Agreed. But I'm trying to wrap this up now, perhaps this should be a separate effort?

nohal commented:
`DimeControl()` behaves differently on different platforms as the native controls wxWidgets uses behave differently and there is not much we can do about it.

transmitterdan commented:
I agree it's a tough problem.  But one thing DimeControl() does that I don't understand is on Windows it just sets
```
window_back_color = wxNullColour;
```
in Day mode. Seems counterintuitive.

rgleason commented:
Don't know if this will help, but here goes.
[Application Wide Theming.docx](https://github.com/user-attachments/files/18347126/Application.Wide.Theming.docx)

transmitterdan commented:
@leamas, because the add connections dialog does not exist at the time the options dialog is shown it does not get DimeControl() treatment. Adding a DimeControl call before ShowModal() will do it. Something like this:
```
    ConnectionEditDialog dialog(this);
    dialog.SetPropsLabel(_("Configure new connection"));
    dialog.SetDefaultConnectionParams();
    DimeControl(&dialog); <----------------------------------
    auto rv = dialog.ShowModal();
```
Granted, it's not optimal, but it's at least consistent with the rest of the options dialog color scheme. And if we ever work on DimeControl this dialog won't get left behind.

bdbcat commented:
Not following closely, but....
In current github master, it is not possible to edit a connection. Edit dialog appears, changes can be made.  But the changes are not applied nor persisted on closing the edit dialog.
Possibly fixed by this PR, not tested.   But we sure could use merging of this PR if it also fixes the edit fault.

leamas commented:
It certainly fixed this. No time right now, it needs some polish before merging

Hakansv commented:
I agree about the not functional edit in master. One have to delete and add a new connection.
But the edit dialog in this PR on Win is not as is in master. It's not sized according to Win scale as Dave have worked with the last years.
On this PC as of my picture above, so not useful. On another is the OK-Cancel not viewed, it's far below the screen bottom while the dialog top is on screen upper edge, so changes cant be saved.

leamas commented:
Edit dialog: Filed #4300

leamas commented:
New upload. Fixed background colours + various sizing issues.

> But the edit dialog in this PR on Win is not as is in master. It's not sized according to Win scale as Dave have worked with the last years.

As I said, the dialog is the same. Furthermore, the scaling is just  a band aid for the fundamentally broken design of the dialog which makes it necessary to set a fixed size rather than to let wxWidgets do the work.-

leamas commented:
Yes, I was bit short there. Still, having a fixed size is IMHO a proof that something is fundamentally wrong, as is the fact that it's not resizeable.

Hakansv commented:
On W10 scaled 175%:
Much better!
Options->Connections:
See shot. 
1. A vertical ruler is there. 
2. Column "Delete" not visible.
3. "Edit" covers all as on master code, e.g. a static size I suppose?
4. "Add new connection" dialog is minimized and not useful. Would benefit a resize function as you mention. I though like this button more than the "+"-sign.
5. Enable/Disable need first a click to select the row. The second click toggles the function. A user may confuse this to it need two clicks to toggle? Once the row is "selected" one click is enough to toggle.
6. Personal view: The options below "General" would be: "[] FilterNMEA... and "Adjust communication pri..." De others may fit below "Advanced".
7. So far so good. I'll later test also on another Win scaled to 100%

![bild](https://github.com/user-attachments/assets/5de925b5-5e0f-40ce-8e6b-4aa865d1a939)

leamas commented:
> Column "Delete" not visible.

Obviously, this is at least partly about a hidden, wide Port column. What is the content of this if you move the scrollbar?

Hakansv commented:
> What is the content of this if you move the scrollbar?

We scroll down like this:
![bild](https://github.com/user-attachments/assets/718d800a-697b-41a1-9f15-b499a42245a6)

rgleason commented:
Heads up.

For Windows, High DPI screens, the recommended scale is in many cases, 250%, if the menu is working nicely for 125% it will not work for 250%!!!
and be unuseable.

Hakansv commented:
Also was all connection user comments set to "0" after that crash. Hmm!

Hakansv commented:
Hint. Make a filter in the text box for user comments so that these characters are not available to use: ";" "|"
The reason may be obvious looking at the config file where these are delimiters for DataConnections=

leamas commented:
New upload: Fixes:

  - Resize polish
  - Make minimum width somewhat bigger to accommodate for larger description fields
  -  Use same (fixed) dialog size  when creating new connection as when using Edit.
  - Don't require double clicks
  - Fix crash when selecting language.

Notes:
1. The crash was a good catch but a bad hint. It was a race condition unrelated to the contents.
2.
> Personal view: The options below "General" would be: "[] FilterNMEA... and "Adjust communication pri..." De others may fit below "Advanced".

No. Less is more. I doubt any new user would have an intuitive understanding what this is all about, and in any case she should be able to get it up & running without it. Hence: advanced.

leamas commented:
The builds fails on two tests. The same tests runs OK on circleci, so something about the CI environment

leamas commented:
@bdbcat : Is there actually a use case for using the garmin protocol carrying Furuno formatted uploads?

leamas commented:
@Hakansv 
> Also was all connection user comments set to "0" after that crash. Hmm!

Well, this is about the broken Edit dialog which does not save it's values.

bdbcat commented:
"Is there actually a use case for using the garmin protocol carrying Furuno formatted uploads"
No.  Completely orthogonal.

transmitterdan commented:
> The builds fails on two tests. The same tests runs OK on circleci, so something about the CI environment

Maybe not, the failure appears to be around lack of opengl in the test environment.  I thought the test cases did not try to show any graphics.

leamas commented:
@transmitterdan 

> Maybe not, the failure appears to be around lack of opengl in the test environment. I thought the test cases did not try to show any graphics.

What makes you think so? I just see an aborted subprocess, aborted for unknown reasons?

leamas commented:
Adding a bunch of new commits.

One is about merging the two checkboxes for "Garmin uploads" and "Furuno uploads" to a single choice, see @bdbcat [above](https://github.com/OpenCPN/OpenCPN/pull/4283#issuecomment-2585275907).

The rest is about new ways to invoke the NMEA debugger. The PR removes this possibility from the connection configuration pages for good reasons, it just does not belong there. But this makes it necessary to create new ways to invoke it. The rest of the commits enables no less than three ways:
  - A new Tools | NMEA Debugger menu entry (requires enabling the menu bar).
  - A new Debug submenu in the canvas root menu. This also need to be enabled in the UI settings.
  - A new hotkey 'E'. Would certainly be happy to use some otfher key, but there are not  that many unused...

Hakansv commented:
I've built the last commits on W10.
Not good at all. I've some general UI comments but first:

1. The connection dialog don't cover all connections listed in the config file. I've 13 connections, only 6 are listed.
2. All connections are disabled on O start,
3. All user comments are replaced by "0" at O start.
4. If I enable a connection it's working until O reboot.
5. I I add a user comment it's there until O reboot.
6. The dialog is not scaled to fit. See pict.

![bild](https://github.com/user-attachments/assets/df8fdfa1-f4bc-4128-b856-1186ed8f6f3d)

transmitterdan commented:
Here is a suggestion:
```
  /** Set up column attributes: alignment, font size, read-only, etc. */
  void SetColAttributes(wxWindow* parent) {
    if (IsWindows()) SetDefaultCellBackgroundColour(GetGlobalColor("DILG1"));  // Set all cells to global color scheme
```
By setting the default background color you can avoid the dark region in the unused cells.  Also, this will avoid the need to set each cell individually unless you want something other than "DILG1" as the background color. And it avoids a ton of wxAsserts in DEBUG builds. It may be better to do this when the wxGrid is created but it works putting it here.
![image](https://github.com/user-attachments/assets/9536b926-2ce0-4235-a1bb-34e296c0e06b)

Hakansv commented:
> I add a user comment it's there until O reboot.

Quick debug during O start.
He reads the config including the user comment and the rest (ConnectionParams)
Very short thereafter he write to config but not using what's read but something else. (TheConnectionParams)

leamas commented:
> Quick debug during O start.
He reads the config including the user comment and the rest (ConnectionParams)
Very short thereafter he write to config but not using what's read but something else. (TheConnectionParams)
 
As I see it, "he" in this context is the broken Edit Connection dialog. This is #4300, not related to this PR. 

Note the only changes in the Edit Connection dialog is that some items are removed together with a now unused and somewhat horrible sort routine. Besides that, it's only about sizing it bigger.

leamas commented:
@transmitterdan 
> Here is a suggestion:

Thanks, seems reasonable will do.

The black background comes after applying DimeControl in options.cpp which creates the whole thing. As you have stated before, this looks like a bug to me. Could it serve any purpose to set the default background colour to effectively be black?

transmitterdan commented:
Black background makes hard to read text that is not almost white.  The DILG1 color seems right visually to me.
DimeControl() should not affect grid background on Windows.  It may do on Linux. But it should be possible to ascertain the background color DimeControl() uses on non-Windows wxGrid objects and replace DILG1 with that color for non-Windows builds. When I get time, I'll try to research that.

leamas commented:
@transmitterdan : thanks

BTW: In your build above, do you have a working vertical scrollbar if you add connections so there at least eight of them?

transmitterdan commented:
I added several connections.  The vertical scroll bar is present, but it does not adjust its range based on newly added connections. Therefore, cannot scroll down to see connections added after 6th one (6th one only partially visible).
![image](https://github.com/user-attachments/assets/3a5c564a-0577-41c4-808b-bfa747fe79da)

leamas commented:
New upload.
  - Fixed an annoying bug caused by accidentally overriding a wxWidgets function.
  - Removed the DimeControl call, from now considered this as buggy on windows. Set the background as proposed by Dan
  - Works for me on Linux and  Windows.

Hakansv commented:
> Note that the only changes in the Edit Connection dialog is that some items are removed together with a now unused and somewhat horrible sort routine. Besides that, it's only about sizing it bigger.

Could be but still if I run master code those enabled before a O reboot are still enabled after the restart.
Running the code in this PR all connections are disabled after a restart.

The broken user comments is valid both in master and here, agree. So If I add a comment using this PR it's not reused after a reboot.

transmitterdan commented:
@Hakansv I can run this same code on Windows and it does not disable all connections. It remembers which are disable/enable ok. I suspect there is something wrong with your opencpn.ini file that may be unrelated to this particular code.  Perhaps there are other settings that end up defaulted? Or perhaps some bogus text appears in or near the communications settings that confuse the ini file parser.

Hakansv commented:
@transmitterdan 
Could be. My config file was the same I've "always" used. But during testing this different versions of this PR something may have happened. I'll redo all connections and test. When it's working here I'll test on some other PCs that haven't seen this code so it's backward compatible. I'll be back...

Hakansv commented:
So I deleted all connections from the config file and redone them from the connections window.
I started with eight connections and then it was the same again. None enabled and user comments replaced by "0" at reboot.
During the process to feed new connections, I gradually noticed that the tool tips for user comments showed the comment from the line above?

Did you tried several connections?

Hakansv commented:
Some of the N0183 TCP connections use the address: "signalk.stupan.se"  and not like 192.168.1.78
Could that disturb the new code?

transmitterdan commented:
Maybe. I'll try something like that and report back.

Hakansv commented:
A for me strange observation after last O closure:

The shutdown is taking an unreasonably long time. The log file may show that something is still "alive" when the main process makes the final note?

```
23:30:51.601 MESSAGE TCDataSource.cpp:41 UnLoading Tide/Current data source: C:\Program Files (x86)\OpenCPN\tcdata\HARMONICS_NO_US.IDX
23:30:51.603 MESSAGE TCDataSource.cpp:41 UnLoading Tide/Current data source: C:\Users\HAS-WIN\Documents\Charts\WP_Marks_Zones_OCPN\Tide Harmonics_old\Europe_tides\HARMONIC.IDX
23:30:51.603 MESSAGE TCDataSource.cpp:41 UnLoading Tide/Current data source: C:\Users\HAS-WIN\Documents\Charts\WP_Marks_Zones_OCPN\Tide Harmonics_old\Europe_tides\EU_current\NL_Current.IDX
23:30:51.610 MESSAGE ocpn_app.cpp:2032 opencpn::MyApp exiting cleanly...

23:30:51.610   ERROR gps_wx_logging.h:14 GPS_Packet_Read: Timeout.  No data received.
23:30:51.610   ERROR gps_wx_logging.h:14 GPS_Get_Ack error -7:
23:30:51.610   ERROR gps_wx_logging.h:14 GPS_Packet_Read: Timeout.  No data received.
23:30:51.610   ERROR gps_wx_logging.h:14 GPS_Get_Ack error: -7
23:30:51.610   ERROR gps_wx_logging.h:14 GPS_Packet_Read: Timeout.  No data received.
23:30:51.611   ERROR gps_wx_logging.h:14 GPS_Get_Ack error -7:

```

Hakansv commented:
Now I have restored my Win config file to normal functionality.
That is using a master build from 2024-12-06 where all connection handle did work. Also edit.
I now have saved a copy of that opencpn.ini
@bdbcat & @leamas 
I think we have to repair the master code before further handling of this PR so we know that is working when later testing updates on this. 
I mean, make edit connections to work again on master branch.

leamas commented:
@Hakansv 
> I think we have to repair the master code before further handling of this PR s

No. In a situation like this it's better to make things feature complete and iron out bugs after that.

Still basically out of time.  If you do have some time, it would be helpful if you could run a bisect to figure out where the Edit dialog went wrong.  If you do, please report results in #4300.

Hakansv commented:
It's possible I can search for the edit bug, I'll see tonight.
Meanwhile:
About the new features in this PR:
If we are now going to change the handling of connections, it should be to something that is adapted to the screens it is displayed on. On all the operating systems we support, there are different ways to scale the screen. This causes problems with static sizes. All dialog boxes and their fonts should be scaled to the current screen. 
Today's "Connections" actually works well but has problems with the display on different systems. Android has everything from a small phone to large tablets. Windows different resolution and scaling from 100 to 250%. I think MacOS has similar development? That is what we have to adapt to in the first place. Delivering a development that involves static size of dialogues can be seen as quite outdated, backward-looking.

leamas commented:
> Delivering a development that involves static size of dialogues can be seen as quite outdated, backward-looking.

Agreed, and this is not.

rgleason commented:
I think Hakan has hit the nail on the head.
The code that handles the screen scaling problem for all supported OS could be used universally, for Opencpn and Plugins, ideally all programmers would use it, as it would solve a big headache for many devs.

> If we are now going to change the handling of connections, it should be to something that is adapted to the screens it is displayed on. On all the operating systems we support, there are different ways to scale the screen. This causes problems with static sizes. All dialog boxes and their fonts should be scaled to the current screen.

leamas commented:
This is no longer a test shot, filing #4307. Closing
